oauth2_http/java/com/google/auth/oauth2/AwsCredentials.java

```diff
@@ -66,17 +71,50 @@ public class AwsCredentials extends ExternalAccountCredentials {
   static final String AWS_SECRET_ACCESS_KEY = "AWS_SECRET_ACCESS_KEY";
   static final String AWS_SESSION_TOKEN = "AWS_SESSION_TOKEN";
 
+  static final String DEFAULT_REGIONAL_CREDENTIAL_VERIFICATION_URL =
+      "https://sts.{region}.amazonaws.com?Action=GetCallerIdentity&Version=2011-06-15";
+
   static final String AWS_IMDSV2_SESSION_TOKEN_HEADER = "x-aws-ec2-metadata-token";
   static final String AWS_IMDSV2_SESSION_TOKEN_TTL_HEADER = "x-aws-ec2-metadata-token-ttl-seconds";
   static final String AWS_IMDSV2_SESSION_TOKEN_TTL = "300";
+
+  static final String AWS_METRICS_HEADER_VALUE = "aws";
+
   private static final long serialVersionUID = -3670131891574618105L;
 
-  private final AwsCredentialSource awsCredentialSource;
+  @Nullable private final AwsCredentialSource awsCredentialSource;
+  @Nullable private final Supplier<AwsSecurityCredentials> awsSecurityCredentialsSupplier;
+  @Nullable private final String regionalCredentialVerificationUrlOverride;
+  @Nullable private final String region;
 
   /** Internal constructor. See {@link AwsCredentials.Builder}. */
   AwsCredentials(Builder builder) {
     super(builder);
-    this.awsCredentialSource = (AwsCredentialSource) builder.credentialSource;
+    // Check that one and only one of supplier or credential source are provided.
+    if (builder.awsSecurityCredentialsSupplier != null && builder.credentialSource != null) {
```

Comment:
This and the rest of the if-else changes indicate, that you can split it. 

It seems to be cleaner to have a separate AwsProgrammaticCredential that extends this one. It can add its own source type, but hide setting the other one. And it can override whatever methods you change here. 

It is easier to understand, could this be an option?

Reply:
Similarly to IdentityPool.

Reply:
Hey Timur, we appreciate your suggestions but we are not going to add new classes. The design has been approved and it is too late to make changes.

Reply:
Discussed offline, we've agreed to go with @westarle's suggestions. Thanks guys.